src/main/java/com/celarli/commons/vfs/provider/google/GCSFileObject.java

```
@@ -319,4 +331,191 @@ protected long doGetLastModifiedTime() throws Exception {
         return super.doGetLastModifiedTime();
     }
 
+
+    @Override
+    public void copyFrom(FileObject file, FileSelector selector) throws FileSystemException {
+
+        this.copyFrom(file, selector, null);
+    }
+
+
+    /**
+     * This method help to copy blob server side if source and destination location belongs to same project.
+     * It also takes listener to report progress of file/folder being copied.
+     *
+     * @param file
+     * @param selector
+     * @param copyStreamListener
+     * @throws FileSystemException
+     */
+    public void copyFrom(FileObject file, FileSelector selector, CopyStreamListener copyStreamListener)
+            throws FileSystemException {
+
+        if (!file.exists()) {
+            throw new FileSystemException("vfs.provider/copy-missing-file.error", file);
+        }
+
+        if (canCopyServerSide(file)) {
+            URLFileName urlFileName = (URLFileName) this.getName();
+            String path = urlFileName.getPath();
+            if (!path.equals("/") && path.startsWith("/")) {
+                path = path.substring(1);
+            }
+
+            String bucket = urlFileName.getHostName();
+
+            GCSFileObject gcsFile = (GCSFileObject) file;
+
+            CopyWriter copyWriter = gcsFile.currentBlob.copyTo(BlobId.of(bucket, path));
+
+            try {
+                //Need to reset file type after copy operation
+                this.injectType(this.doGetType());
+            }
+            catch (Exception e) {
+                e.printStackTrace();
```

either swallow the exception or log it or rethrow it